src/sys/windows/named_pipe.rs

```diff
@@ -0,0 +1,610 @@
+//! to a named pipe. Afterwards you can use the `into_raw_handle` method coupled
+//! with the `NamedPipe::from_raw_handle` method to convert that to a named pipe
+//! that can operate asynchronously. Don't forget to pass the
+//! `FILE_FLAG_OVERLAPPED` flag when opening the `File`.
+
+use crate::{poll, Registry};
+use crate::sys::windows::Overlapped;
+
+use std::ffi::OsStr;
+use std::fmt;
+use std::io::{self, Read};
+use std::mem;
+use std::os::windows::io::*;
+use std::slice;
+use std::sync::atomic::{AtomicUsize, AtomicBool};
+use std::sync::atomic::Ordering::{Relaxed, SeqCst};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+// use mio::windows;
+// use mio::{Evented, Poll, PollOpt, Ready, Registration, SetReadiness, Token};
+use miow::iocp::CompletionStatus;
+use miow::pipe;
+use winapi::shared::winerror::*;
+use winapi::um::ioapiset::*;
+use winapi::um::minwinbase::*;
+
+macro_rules! offset_of {
+    ($t:ty, $($field:ident).+) => (
+        &(*(0 as *const $t)).$($field).+ as *const _ as usize
+    )
+}
+
+macro_rules! overlapped2arc {
+    ($e:expr, $t:ty, $($field:ident).+) => ({
+        let offset = offset_of!($t, $($field).+);
+        debug_assert!(offset < mem::size_of::<$t>());
+        Arc::from_raw(($e as usize - offset) as *mut $t)
+    })
+}
+
+fn would_block() -> io::Error {
+    io::ErrorKind::WouldBlock.into()
+}
+
+/// Representation of a named pipe on Windows.
+///
+/// This structure internally contains a `HANDLE` which represents the named
+/// pipe, and also maintains state associated with the mio event loop and active
+/// I/O operations that have been scheduled to translate IOCP to a readiness
+/// model.
+pub struct NamedPipe {
+    inner: Arc<Inner>,
+}
+
+struct Inner {
+    handle: pipe::NamedPipe,
+
+    connect: Overlapped,
+    connecting: AtomicBool,
+
+    read: Overlapped,
+    write: Overlapped,
+
+    io: Mutex<Io>,
+
+    pool: Mutex<BufferPool>,
+}
+
+struct Io {
+    read: State,
+    read_waker: Option<Waker>,
+    write: State,
+    write_waker: Option<Waker>,
+    connect_error: Option<io::Error>,
+}
+
+#[derive(Debug)]
+enum State {
+    None,
+    Pending(Vec<u8>, usize),
+    Ok(Vec<u8>, usize),
+    Err(io::Error),
+}
+
+// Odd tokens are for named pipes
+static NEXT_TOKEN: AtomicUsize = AtomicUsize::new(1);
+
+fn _assert_kinds() {
+    fn _assert_send<T: Send>() {}
+    fn _assert_sync<T: Sync>() {}
+    _assert_send::<NamedPipe>();
+    _assert_sync::<NamedPipe>();
+}
+
+impl NamedPipe {
+    /// Creates a new named pipe at the specified `addr` given a "reasonable
+    /// set" of initial configuration options.
+    pub fn new<A: AsRef<OsStr>>(
+        addr: A,
+        registry: &Registry,
+    ) -> io::Result<NamedPipe> {
+        let pipe = pipe::NamedPipe::new(addr)?;
+        NamedPipe::from_raw_handle(pipe.into_raw_handle(), registry)
+    }
+
+    /// TODO: Dox
+    pub fn from_raw_handle(
```

Comment:
This must be `unsafe` because it calls [`FromRawHandle::from_raw_handle`](https://doc.rust-lang.org/std/os/windows/io/trait.FromRawHandle.html) on `pipe::NamedPipe`, the contract of which the caller must uphold.